src/Functions/UserDefined/UserDefinedSQLFunctionVisitor.cpp

```diff
@@ -19,24 +22,96 @@ namespace ErrorCodes
     extern const int UNSUPPORTED_METHOD;
 }
 
-void UserDefinedSQLFunctionMatcher::visit(ASTPtr & ast, Data &)
+void UserDefinedSQLFunctionVisitor::visit(ASTPtr & ast)
 {
-    auto * function = ast->as<ASTFunction>();
-    if (!function)
+    const auto visit_child_with_shared_ptr = [&](ASTPtr & child)
+    {
+        if (!child)
+            return;
+
+        auto * old_value = child.get();
+        visit(child);
+        ast->setOrReplace(old_value, child);
```

Comment:
I don't like that
* We use dynamic_cast in `setOrReplace` - it is not needed. Maybe write a simple for-loop here?
* We iterate all children every time unconditionally. Probably check `old_value != child.get()`?